storage/collections.go

```diff
@@ -35,30 +37,21 @@ type Collections interface {
 
 	// Store inserts the collection keyed by ID and all constituent
 	// transactions.
+	// This is used by execution node storing collections
+	// No errors are expected during normal operation.
 	// No errors are expected during normal operation.
-	Store(collection *flow.Collection) error
+	Store(collection *flow.Collection) (flow.LightCollection, error)
 
 	// Remove removes the collection and all constituent transactions.
 	// No errors are expected during normal operation.
 	Remove(collID flow.Identifier) error
 
-	// StoreLightAndIndexByTransaction inserts the light collection (only
-	// transaction IDs) and adds a transaction id index for each of the
-	// transactions within the collection (transaction_id->collection_id).
-	//
-	// NOTE: Currently it is possible in rare circumstances for two collections
-	// to be guaranteed which both contain the same transaction (see https://github.com/dapperlabs/flow-go/issues/3556).
-	// The second of these will revert upon reaching the execution node, so
-	// this doesn't impact the execution state, but it can result in the Access
-	// node processing two collections which both contain the same transaction (see https://github.com/dapperlabs/flow-go/issues/5337).
-	// To handle this, we skip indexing the affected transaction when inserting
-	// the transaction_id->collection_id index when an index for the transaction
-	// already exists.
-	//
+	// StoreAndIndexByTransaction stores the collection and indexes it by transaction.
+	// This is used by access node storing collections for finalized blocks
 	// No errors are expected during normal operation.
-	StoreLightAndIndexByTransaction(collection *flow.LightCollection) error
+	StoreAndIndexByTransaction(lctx lockctx.Proof, collection *flow.Collection) (flow.LightCollection, error)
```

Comment:
In my opinion, a lock proof is needed if and only if the operations are added to an externally provided write batch. That's why I objected to [`LightCollectionsStore.Persist(batch storage.ReaderBatchWriter) error`](https://github.com/onflow/flow-go/blob/51a91beb17911af0c7eba5133c47a90c38eb7dda/module/executiondatasync/optimistic_sync/persisters/stores/light_collections.go#L33), because atomicity with the external commit of the provided `batch` is not maintained. Here, we have the reverse situation: `StoreAndIndexByTransaction` internally does its own atomic read and write? There isn't anything to synchronize with the external caller, because by the time `StoreAndIndexByTransaction` returns, the write is already committed

```suggestion
	StoreAndIndexByTransaction(lcollection *flow.Collection) (flow.LightCollection, error)
```

In comparison, for the [`BatchStoreAndIndexByTransaction`](https://github.com/onflow/flow-go/blob/51a91beb17911af0c7eba5133c47a90c38eb7dda/storage/collections.go#L55) below, we need to pass a lock, because we are also passing a `ReaderBatchWriter`

Reply:
>In my opinion, a lock proof is needed if and only if the operations are added to an externally provided write batch.

I don't think a lock proof has to be tired with a write batch. I think the a lock proof is needed to sync with other callers that is writing to the same data. 

In our case here, both `StoreAndIndexByTransaction`, and `BatchStoreAndIndexByTransaction` will write the same data, the lock is to prevent dirty reads/writes in these calls.

>because by the time StoreAndIndexByTransaction returns, the write is already committed

I'm not sure what you mean. Do you mean we don't need a lock here? I think the locking is required, because otherwise the reads involved in this call could be dirty.

Reply:
Sorry, I added a long comment but not really made my point clear: **I think we don't need a lock here**. However, this is not critical, since locking to much only has performance penalties. 

For the sake of time efficiency, it's probably best to discuss this in person. I added this as medium priority point to issue https://github.com/onflow/flow-go/issues/7682 (medium prio, because I think it is relativeluy important that we are aligned on best practises wrt locking design patterns.)